### PR TITLE
PHOENIX-6583 Inserting explicit Null into a (fixed length) binary field is stored as an array of zeroes

### PR DESCRIPTION
**Pull request by alejandro-anadon**
PHOENIX-6583 Inserting explicit Null into a (fixed length) binary field is stored as an array of zeroes

In the test BinaryTypeIT:
1) it  create the table
2) it make a simple upsert without using the binary field and test if it is null.
3)it makes an upsert to the binary field but with a null value. Without the change, it stores a 0x00 array of bytes instead of a null.
4) it makes an upsert to the binary field but now with a value.
5)it makes an upsert to the binary field again with a null. Without the change, it does not remove the data and it is not a null value.

IMPORTANT: while I was doing the tests, I tried to pass all the test, but failed in the test "org.apache.phoenix.end2end.DefaultColumnValueIT.testDefaultAllDataTypesKeyValueCol" , but in my opinion the fail is regarding what I have explained it in JIRA PHOENIX-6585 report. This is the stack trace of the test:

[ERROR] Tests run: 24, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 230.567 s <<< FAILURE! - in org.apache.phoenix.end2end.DefaultColumnValueIT
[ERROR] org.apache.phoenix.end2end.DefaultColumnValueIT.testDefaultAllDataTypesKeyValueCol  Time elapsed: 4.806 s  <<< FAILURE!
arrays first differed at element [4]; expected:<0> but was:<2>
        at org.junit.internal.ComparisonCriteria.arrayEquals(ComparisonCriteria.java:78)
        at org.junit.internal.ComparisonCriteria.arrayEquals(ComparisonCriteria.java:28)
        at org.junit.Assert.internalArrayEquals(Assert.java:534)
        at org.junit.Assert.assertArrayEquals(Assert.java:343)
        at org.junit.Assert.assertArrayEquals(Assert.java:354)
        at org.apache.phoenix.end2end.DefaultColumnValueIT.testDefaultAllDataTypes(DefaultColumnValueIT.java:581)
        at org.apache.phoenix.end2end.DefaultColumnValueIT.testDefaultAllDataTypesKeyValueCol(DefaultColumnValueIT.java:506)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
        at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
        at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunne

### CODE REVIEW

**Review by stoty — CHANGES_REQUESTED** [review]
The failing org.apache.phoenix.end2end.DefaultColumnValueIT.testDefaultAllDataTypesPrimaryKey()
test shows that the patch breaks existing functionality.

PHOENIX-6585 is about non PK fields, while that test checks PK fields, so it does not apply.

Please also fix the formatting issues shown by the CI.

**Comment by alejandro-anadon** [comment]
Hi,

I think that I solved. I forget to check ptr. Before I was checking only for the Object. In my test, now is not having failures.

But now, I found something that for me is streange. I tested it not only with fixed binaries, but algo with integers. At the begining I thought that it was bacause my fixed; so I proved without my changes. But I had the same result, so I thing that it is not because of my changes and it comes before.

The question is that it seems that there is a conflict between "field is null" and "filed=null". It is supoused that the result should be the same; but it is not. Followin is what I mean (you can try in master branch without my changes):

0: jdbc:phoenix:> create table  DUMMY (id integer not null, int2 INTEGER, text varchar(255), testbin binary(16), CONSTRAINT pk primary key (id));
No rows affected (2.478 seconds)


0: jdbc:phoenix:> upsert into dummy values (1);
1 row affected (0.055 seconds)

0: jdbc:phoenix:> select * from dummy;
+----+------+------+---------+
| ID | INT2 | TEXT | TESTBIN |
+----+------+------+---------+
| 1  | null |      |         |
+----+------+------+---------+
1 row selected (0.039 seconds)

0: jdbc:phoenix:> select * from dummy where int2 is null;
+----+------+------+---------+
| ID | INT2 | TEXT | TESTBIN |
+----+------+------+---------+
| 1  | null |      |         |
+----+------+------+---------+
1 row selected (0.03 seconds)

0: jdbc:phoenix:> select * from dummy where int2=null;
+----+------+------+---------+
| ID | INT2 | TEXT | TESTBIN |
+----+------+------+---------+
+----+------+------+---------+
No rows selected (0.033 seconds)


"select * from dummy where int2 is null;" should returns the same as "select * from dummy where int2=null;". should't ?

**Comment by stoty** [comment]
The null behaviour you describe is correct, as defined by SQL spec. Look it up.
Please fix the formatting problems reported by Yetus.

**Comment by alejandro-anadon** [comment]
Done.
About SQL spec, I do not know why it was speced like that , but I do not agree (for me it has no sense, but I tested it postgres and the  behaviour is the same)... but  it's like fighting giants.


NOTE: I am always getting the same error report in test, both before and after the changes I made in PhoenixMetricsIT.testGlobalPhoenixMetricsForQueries. This is the stack trace:

[INFO] Running org.apache.phoenix.monitoring.PhoenixTableLevelMetricsIT
[ERROR] Tests run: 24, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 289.222 s <<< FAILURE! - in org.apache.phoenix.monitoring.PhoenixMetricsIT
[ERROR] org.apache.phoenix.monitoring.PhoenixMetricsIT.testGlobalPhoenixMetricsForQueries  Time elapsed: 7.059 s  <<< FAILURE!
java.lang.AssertionError
        at org.junit.Assert.fail(Assert.java:87)
        at org.junit.Assert.assertTrue(Assert.java:42)
        at org.junit.Assert.assertTrue(Assert.java:53)
        at org.apache.phoenix.monitoring.PhoenixMetricsIT.testGlobalPhoenixMetricsForQueries(PhoenixMetricsIT.java:199)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
        at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
        at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
        at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
        at org.apache.phoenix.SystemExitRule$1.evaluate(SystemExitRule.java:40)
        at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
        at org.junit.runners.Suite.runChild(Suite.java:128)
        at org.junit.runners.Suite.runChild(Suite.java:27)
        at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
        at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
        at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
        at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)
        at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)
        at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:107)
        at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:83)
        at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
        at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:157)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
        at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
        at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)

**Comment by alejandro-anadon** [comment]
(I apologies if this is not the correct place to say following, but I think that if somebody read this conversation may get the the same doubt that I had about regarding SQL spec)
I read the following article I clarified my ideas:
https://www.xaprb.com/blog/2006/05/18/why-null-never-compares-false-to-anything-in-sql/

It remenber me a puzzle that I read in this book (http://www.javapuzzlers.com/) where you can have a instance like this:
---
if (a!=a)
    System.out.println('Hello!! I am here");
---
And the mesage will be shown. How? well... initializating a like:
double a = Math.log(-1);

In this case, a is 'NaN'; and java spec saiys that 'NaN' is not equal to anything... neither NaN, so ' NaN  != NaN' is true.

**Comment by alejandro-anadon** [comment]
sorry for hightlighting the parragraph about the book. It was not my intent. I don't know why it did it... I only write simple text... next time I will "preview" the mesage. This is my first time using github

**Comment by alejandro-anadon** [comment]
Hi,
I do not understand the error that Yetus(jenkins) is reporting because the file phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java was deleted from the branck and it does not exists. You can check it in the PHOENIX-6583 branch on alejandro-anadon/phoenix. 

Here is when I deletted it:
https://github.com/apache/phoenix/pull/1342/commits/eae4b115e18f932debb6aca9ce1b6d8a7ff5c2df

Just as you suggested, that test was moved to phoenix-core/src/it/java/org/apache/phoenix/end2end/NullIT.java.

In my enviroment, the only failed test is PhoenixMetricsIT.testGlobalPhoenixMetricsForQueries; but I passed also all the test before my changes, and I have the same failed test; so I guess it has nothing to do with my changes that I have made.

I am going to post ascreen log file in case it may help. It is absolutly from begining. I mean, from firt 'git clone' of my repositorie.

(I also have an error after all the test has been done. It is  another error when it is supoused to begin to generate the report when it try to execute' org.jacoco:jacoco-maven-plugin:0.8.7:report' and I do not know why. But It has nothing of relation with , and I do not care too much because if I try to build the binaries without test, it works until the end.

(I am trying to build with hbase profile 2.1.9 and using -Dmaven.compiler.executable=/usr/lib/jvm/jdk1.8.0_202/bin/javac  because I have by default an openjdk version "11.0.12" 2021-07-20).

**Comment by alejandro-anadon** [comment]
Here is the file log
[PHOENIX-6583-SCREEN-LOG-IN-BUILDING.txt](https://github.com/apache/phoenix/files/7499097/PHOENIX-6583-SCREEN-LOG-IN-BUILDING.txt)

File: phoenix-core/src/main/java/org/apache/phoenix/schema/types/PBinary.java [hunk 1]
```diff
@@ -36,11 +36,13 @@ private PBinary() {
...
-            Integer actualScale, SortOrder actualModifier, Integer desiredMaxLength, Integer desiredScale,
-            SortOrder expectedModifier) {
-        PVarbinary.INSTANCE.coerceBytes(ptr, o, actualType, actualMaxLength, actualScale, actualModifier, desiredMaxLength, desiredScale, expectedModifier);
-        if (null != desiredMaxLength && null != expectedModifier) {
+    public void coerceBytes(ImmutableBytesWritable ptr, Object o, PDataType actualType,
+            Integer actualMaxLength, Integer actualScale, SortOrder actualModifier,
+            Integer desiredMaxLength, Integer desiredScale, SortOrder expectedModifier) {
+        PVarbinary.INSTANCE.coerceBytes(ptr, o, actualType, actualMaxLength, actualScale,
+                actualModifier, desiredMaxLength, desiredScale, expectedModifier);
+        if ((o != null || ptr.getLength() > 0)
```
Comment (stoty) [inline phoenix-core/src/main/java/org/apache/phoenix/schema/types/PBinary.java:44]: The `o != null` check seems redundant.

**Review by stoty — CHANGES_REQUESTED** [review]
Please also fix the remaining spaces at the end of lines.

**Comment by alejandro-anadon** [comment]
You are right. With 'ptr.getLength() > 0' is enough.
At the begining I thought that only 'o' should be checked.
Later I added ptr.lenght (it was necesary for default values) without cheking if 'o != null' could be removed.
Finally, as you sugest, I tested  with only  'ptr.getLength() > 0'.
And, of course, if there is not something that I do not see, It can be removed.


About the test, in my enviroment, I tested the changes sereval times. the results are:
default-test -> never fails
ParallelStatsEnabledTest ->  never fails
ParallelStatsDisabledTest -> never fails
NeedTheirOwnClusterTests -> randon failures

Those random test are not allways the same failures.

 for example, in my last test, the failure is in: 'PermissionNSEnabledWithCustomAccessControllerIT>BasePermissionsIT.testSystemTablePermissions:1027->BasePermissionsIT.verifyAllowed:894->BasePermissionsIT.verifyAllowed:901 » UndeclaredThrowable'.

But other time, I have 'PhoenixMetricsIT.testGlobalPhoenixMetricsForQueries:199' (now I don't know if that error was with jdk8 or jdk11)

Another time I had:
'MaxConcurrentConnectionsIT.testDeleteRuntimeFailureClosesConnections:124 Found 1 connections still open. expected:<0> but was:<1>'
and 'LogicalTableNameIT.testUpdatePhysicalIndexTableName_runScrutiny:215->LogicalTableNameBaseIT.test_IndexTableChange:187 » SnapshotCreation'.

Summing up, in 'NeedTheirOwnClusterTests', I can't give a fault pattern, but I think the faults I have have nothing to do with the change I am proposing.

Finally, when maven is going to build the report with jacoco, it allways fails (in my enviroment).

[ERROR] Failed to execute goal org.jacoco:jacoco-maven-plugin:0.8.7:report (report) on project phoenix-core: Execution report of goal org.jacoco:jacoco-maven-plugin:0.8.7:report failed: Incompatible execution data for class org/apache/hadoop/hbase/regionserver/MetricsRegionAggregateSourceImpl with id 2687434e0753a2bc. -> [Help 1]

(I drop the screen log of my last execution).

[PHOENIX-6583-SCREEN-LOG-IN-BUILDING.txt](https://github.com/apache/phoenix/files/7507090/PHOENIX-6583-SCREEN-LOG-IN-BUILDING.txt)

I wanted to point out that, regarding test, I had lots of problems and fails when using 'openjdk version "11.0.12" 2021-07-20' with semi-random test failures. So I decided to change all to 'java version "1.8.0_202"'.
Since them, I only have random failures in NeedTheirOwnClusterTests . I followed the instruction in https://github.com/apache/phoenix/blob/master/BUILDING.md . so I sugest, if it is necesary a specific jvm to compile and test, it is advisable to indicate it that file.

Besides, when you ask to 'Please also fix the remaining spaces at the end of lines.', I thing that there is some king of cache with file 'phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java' beacuse it was removed from the branch in the commit eae4b11. I saw several times to "remove spaces' in that file. But the file is removed. You can check it int the PHOENIX-6583 branch on alejandro-anadon/phoenix.
(https://github.com/alejandro-anadon/phoenix).

In the checkstile chekers I see follow (but not allways apears... sometimes appers, another not):

phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:40:        
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:42:        
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:44:        
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:46:            
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:49:        
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:51:        
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:52:            
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:59:            
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:66:            
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:73:            
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:80:           
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:81:        
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:84:    

But that file does not exist anymore. The test that I propose was moved to phoenix-core/src/it/java/org/apache/phoenix/end2end/NullIT.java as you sugested.

**Comment by alejandro-anadon** [comment]
It keeps saying: "Apache Yetus(jenkins) error: blanks — The patch has 13 line(s) that end in blanks. " over the file that does not exits any more in  PHOENIX-6583 branch on alejandro-anadon/phoenix:

phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:40:        
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:42:        
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:44:        
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:46:            
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:49:        
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:51:        
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:52:            
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:59:            
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:66:            
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:73:            
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:80:           
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:81:        
phoenix-core/src/it/java/org/apache/phoenix/end2end/BinaryTypeIT.java:84:        

That file was deleted in the commit https://github.com/apache/phoenix/commit/eae4b115e18f932debb6aca9ce1b6d8a7ff5c2df

**Comment by stoty** [comment]
Thank you @alejandro-anadon.
The blank lines issue is indeed a Yetus problem.
The Jacoco instability is also known, but we do no have a solution for it.
There are also some flaky tests, so that's OK.

**Comment by stoty** [comment]
BTW the solution you found to is te same the code that is used in PChar.

**Review by stoty — APPROVED** [review]

**Comment by stoty** [comment]
@alejandro-anadon 
The patch is good commit, but I have noticed that your email address has an IP address in it.
Do you want to use that as your committer email ?
If not, then please push another commit with the email that you want to use.
If yes, please confirm it.

**Comment by alejandro-anadon** [comment]
Hi,

now I am lost. I saw that you closed the PR with unmerged commits. Now I don't know how to proceed.
Should I  delete alejandro-anadon:PHOENIX-6583 branch, open a new one (same name?) and do a new PR ? (Belive or not... I am a new user of git. I had been using subversion until this colavolarion)

**Comment by stoty** [comment]
I've merged this patch manually, not via the GitHub interface. It's already in the master repo (check it).
You can delete this PR, or leave it as it is, it doesn't matter.

**Pull request closed without merging**